### PR TITLE
Invalid use of @overload

### PR DESCRIPTION
**Pull request by 16234397**
No SDKs were effected.l

### CODE REVIEW

File: tools/pylint-extensions/azure-pylint-guidelines-checker/pylint_guidelines_checker.py [hunk 1]
```diff
@@ -2807,6 +2807,79 @@ def visit_import(self, node):
...
+            "Do not mix async and synchronous overloads",
+            "invalid-use-of-overload",
+            "Functions and their overloads must be either all async or all synchronous.",
+        ),
+    }
+
+    def visit_module(self, node):
+        """Check that use of the @overload decorator matches the async/sync nature of the underlying function"""
+        try:
+            klass = node.parent.parent.parent
```
Comment (l0lawrence) [inline tools/pylint-extensions/azure-pylint-guidelines-checker/pylint_guidelines_checker.py:2826]: Are you trying to access the class here --  Could you maybe use visit_classdef and visit_functiondef? and track the functions/function names with a list that way?

File: tools/pylint-extensions/azure-pylint-guidelines-checker/tests/test_files/invalid_use_of_overload_acceptable.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+@overload
+def double(a: str)  -> Awaitable[int]:
+    ...
+
+@overload
+def double(a: int) -> Awaitable[int]:
+    ...
+
+async def double(a: Union[str, int]) -> int:
+    if isinstance(a, str):
```
Comment (l0lawrence) [inline tools/pylint-extensions/azure-pylint-guidelines-checker/tests/test_files/invalid_use_of_overload_acceptable.py:15]: shouldn't this fail here?

Comment (16234397) [reply tools/pylint-extensions/azure-pylint-guidelines-checker/tests/test_files/invalid_use_of_overload_acceptable.py:15]: My understanding is that because the overloaded functions have an Awaitable[int] as a return type this makes them an async function too. If this isn't the case it does simplify the implementation.

File: tools/pylint-extensions/azure-pylint-guidelines-checker/pylint_guidelines_checker.py (hunk 1 above)
Comment (16234397) [reply tools/pylint-extensions/azure-pylint-guidelines-checker/pylint_guidelines_checker.py]: I have removed some code that wasn't needed anymore, I originally had it looking at a call level.

Do you want me to check at a module level if their are no classes or just change it so that I looks at a class level only?

File: tools/pylint-extensions/azure-pylint-guidelines-checker/tests/test_files/invalid_use_of_overload_acceptable.py (hunk 2 above)
Comment (l0lawrence) [reply tools/pylint-extensions/azure-pylint-guidelines-checker/tests/test_files/invalid_use_of_overload_acceptable.py]: Ah sorry I did not see the awaitable :)

File: tools/pylint-extensions/azure-pylint-guidelines-checker/pylint_guidelines_checker.py (hunk 1 above)
Comment (16234397) [reply tools/pylint-extensions/azure-pylint-guidelines-checker/pylint_guidelines_checker.py]: I just had a look and there are almost 53,000 occurances of @overload, I checked a few and they are all in a class.

File: tools/pylint-extensions/azure-pylint-guidelines-checker/pylint_guidelines_checker.py [hunk 3]
```diff
@@ -2864,13 +2933,6 @@ def visit_import(self, node):
                 )
 
 
-# [Pylint] custom linter check for invalid use of @overload #3229
-# [Pylint] Custom Linter check for Exception Logging #3227
```
[the remark this answers is not in the document]
Comment (16234397) [reply tools/pylint-extensions/azure-pylint-guidelines-checker/pylint_guidelines_checker.py:2868]: Yes, your right, at the time I thought they were only on our copy. I have added them back in :-)

File: tools/pylint-extensions/azure-pylint-guidelines-checker/pylint_guidelines_checker.py (hunk 1 above)
Comment (l0lawrence) [reply tools/pylint-extensions/azure-pylint-guidelines-checker/pylint_guidelines_checker.py]: Looks good to me :)

**Review by l0lawrence — APPROVED** [review]

File: tools/pylint-extensions/azure-pylint-guidelines-checker/tests/test_files/do_not_hardcode_connection_verify_acceptable.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
+class InstanceVariableError:
+    def __init__(self):
+        self.connection_verify = None
+        self.self = self
+
+
+class VariableError:
+    connection_verify = None
```
Comment (l0lawrence) [inline tools/pylint-extensions/azure-pylint-guidelines-checker/tests/test_files/do_not_hardcode_connection_verify_acceptable.py:8]: can we remove these they are not related overload rule :)

File: tools/pylint-extensions/azure-pylint-guidelines-checker/tests/test_files/invalid_use_of_overload_acceptable.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+class testingOverload:
+    @overload
+    def double(a: str)  -> Awaitable[int]:
+        ...
+
+    @overload
+    def double(a: int) -> Awaitable[int]:
+        ...
+
+    async def double(a: Union[str, int]) -> int:
```
Comment (l0lawrence) [inline tools/pylint-extensions/azure-pylint-guidelines-checker/tests/test_files/invalid_use_of_overload_acceptable.py:14]: after further discussion could we remove this scenario from the checker for now, unless you have a use case for this -- it does not seem to appear in our repo

Comment (16234397) [reply tools/pylint-extensions/azure-pylint-guidelines-checker/tests/test_files/invalid_use_of_overload_acceptable.py:14]: I can' remove this out, It was just to make sure that all scenarios were checked.

File: tools/pylint-extensions/azure-pylint-guidelines-checker/tests/test_files/do_not_hardcode_connection_verify_acceptable.py (hunk 4 above)
Comment (16234397) [reply tools/pylint-extensions/azure-pylint-guidelines-checker/tests/test_files/do_not_hardcode_connection_verify_acceptable.py:8]: For our final report we tried to put all of our checkers in one branch so that code coverage reports were easier, this didn't work, it changed our pull requests - because there has been so many changes to the files since I can't revert the pull - I'm going to create a V2 branch with the below removed on a new pull request.
